autofs-5.0.8 - fix map source with type lookup

From: Ian Kent <raven@themaw.net>

If the map source type is specified the map name must be a full
path.

But we should be able to give a plain map name and expect autofs
to look in the compiled in maps directory.
---
 CHANGELOG       |    1 
 daemon/lookup.c |  111 +++++++++++++++++++++++++++++++++++++++++++++++++-------
 2 files changed, 99 insertions(+), 13 deletions(-)

--- autofs-5.0.5.orig/daemon/lookup.c
+++ autofs-5.0.5/daemon/lookup.c
@@ -100,6 +100,27 @@ static int do_read_master(struct master
 	return status;
 }
 
+static char *find_map_path(struct map_source *map)
+{
+	struct stat st;
+	char *path;
+
+	path = malloc(strlen(AUTOFS_MAP_DIR) + strlen(map->argv[0]) + 2);
+	if (!path)
+		return NULL;
+
+	strcpy(path, AUTOFS_MAP_DIR);
+	strcat(path, "/");
+	strcat(path, map->argv[0]);
+
+	if (!stat(path, &st))
+		return path;
+
+	free(path);
+
+	return NULL;
+}
+
 static int read_master_map(struct master *master, char *type, time_t age)
 {
 	unsigned int logopt = master->logopt;
@@ -390,6 +411,42 @@ static void argv_cleanup(void *arg)
 	return;
 }
 
+static int lookup_map_read_map(struct autofs_point *ap,
+			       struct map_source *map, time_t age)
+{
+	char *path;
+
+	if (!map->argv[0])
+		return NSS_STATUS_UNKNOWN;
+
+	/*
+	 * This is only called when map->type != NULL.
+	 * We only need to look for a map if source type is
+	 * file and the map name doesn't begin with a "/".
+	 */
+	if (strncmp(map->type, "file", 4))
+		return do_read_map(ap, map, age);
+
+	if (map->argv[0][0] == '/')
+		return do_read_map(ap, map, age);
+
+	path = find_map_path(map);
+	if (!path)
+		return NSS_STATUS_UNKNOWN;
+
+	if (map->argc >= 1) {
+		if (map->argv[0])
+			free((char *) map->argv[0]);
+		map->argv[0] = path;
+	} else {
+		error(ap->logopt, "invalid arguments for autofs_point");
+		free(path);
+		return NSS_STATUS_UNKNOWN;
+	}
+
+	return do_read_map(ap, map, age);
+}
+
 static enum nsswitch_status read_map_source(struct nss_source *this,
 		struct autofs_point *ap, struct map_source *map, time_t age)
 {
@@ -426,14 +483,10 @@ static enum nsswitch_status read_map_sou
 	tmap.argc = 0;
 	tmap.argv = NULL;
 
-	path = malloc(strlen(AUTOFS_MAP_DIR) + strlen(map->argv[0]) + 2);
+	path = find_map_path(map);
 	if (!path)
 		return NSS_STATUS_UNKNOWN;
 
-	strcpy(path, AUTOFS_MAP_DIR);
-	strcat(path, "/");
-	strcat(path, map->argv[0]);
-
 	if (map->argc >= 1) {
 		tmap.argc = map->argc;
 		tmap.argv = copy_argv(map->argc, map->argv);
@@ -494,7 +547,7 @@ int lookup_nss_read_map(struct autofs_po
 				debug(ap->logopt,
 				      "reading map %s %s",
 				       map->type, map->argv[0]);
-			result = do_read_map(ap, map, age);
+			result = lookup_map_read_map(ap, map, age);
 			map = map->next;
 			continue;
 		}
@@ -756,6 +809,43 @@ static int lookup_name_source_instance(s
 	return do_lookup_mount(ap, instance, name, name_len);
 }
 
+static int do_name_lookup_mount(struct autofs_point *ap,
+				struct map_source *map,
+				const char *name, int name_len)
+{
+	char *path;
+
+	if (!map->argv[0])
+		return NSS_STATUS_UNKNOWN;
+
+	/*
+	 * This is only called when map->type != NULL.
+	 * We only need to look for a map if source type is
+	 * file and the map name doesn't begin with a "/".
+	 */
+	if (strncmp(map->type, "file", 4))
+		return do_lookup_mount(ap, map, name, name_len);
+
+	if (map->argv[0][0] == '/')
+		return do_lookup_mount(ap, map, name, name_len);
+
+	path = find_map_path(map);
+	if (!path)
+		return NSS_STATUS_UNKNOWN;
+
+	if (map->argc >= 1) {
+		if (map->argv[0])
+			free((char *) map->argv[0]);
+		map->argv[0] = path;
+	} else {
+		error(ap->logopt, "invalid arguments for autofs_point");
+		free(path);
+		return NSS_STATUS_UNKNOWN;
+	}
+
+	return do_lookup_mount(ap, map, name, name_len);
+}
+
 static enum nsswitch_status lookup_map_name(struct nss_source *this,
 			struct autofs_point *ap, struct map_source *map,
 			const char *name, int name_len)
@@ -791,14 +881,10 @@ static enum nsswitch_status lookup_map_n
 	tmap.argc = 0;
 	tmap.argv = NULL;
 
-	path = malloc(strlen(AUTOFS_MAP_DIR) + strlen(map->argv[0]) + 2);
+	path = find_map_path(map);
 	if (!path)
 		return NSS_STATUS_UNKNOWN;
 
-	strcpy(path, AUTOFS_MAP_DIR);
-	strcat(path, "/");
-	strcat(path, map->argv[0]);
-
 	if (map->argc >= 1) {
 		tmap.argc = map->argc;
 		tmap.argv = copy_argv(map->argc, map->argv);
@@ -907,8 +993,7 @@ int lookup_nss_mount(struct autofs_point
 		sched_yield();
 
 		if (map->type) {
-			result = do_lookup_mount(ap, map, name, name_len);
-
+			result = do_name_lookup_mount(ap, map, name, name_len);
 			if (result == NSS_STATUS_SUCCESS)
 				break;
 
--- autofs-5.0.5.orig/CHANGELOG
+++ autofs-5.0.5/CHANGELOG
@@ -185,6 +185,7 @@
 - fix use get_proximity() without libtirpc.
 - document allowed map sources in auto.master.
 - fix external env configure.
+- fix map source with type lookup.
 
 03/09/2009 autofs-5.0.5
 -----------------------
